src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.PendingBranchesCollection.cs

```diff
@@ -0,0 +1,134 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal partial class AbstractFlowPass<TLocalState, TLocalFunctionState>
+    {
+        protected sealed class PendingBranchesCollection : IEnumerable<PendingBranch>
+        {
+            private ArrayBuilder<PendingBranch> _branches;
+            private PooledDictionary<LabelSymbol, ArrayBuilder<PendingBranch>>? _branchesToLabel;
+
+            internal PendingBranchesCollection()
+            {
+                _branches = ArrayBuilder<PendingBranch>.GetInstance();
+            }
+
+            internal void Free()
+            {
+                _branches.Free();
+                FreeBranchesToLabel();
+            }
+
+            internal void Clear()
+            {
+                _branches.Clear();
+                FreeBranchesToLabel();
+            }
+
+            private void FreeBranchesToLabel()
+            {
+                if (_branchesToLabel is { })
+                {
+                    foreach (var branches in _branchesToLabel.Values)
+                    {
+                        branches.Free();
+                    }
+                    _branchesToLabel.Free();
+                    _branchesToLabel = null;
+                }
+            }
+
+            internal ImmutableArray<PendingBranch> ToImmutable()
+            {
+                return _branchesToLabel is null ?
+                    _branches.ToImmutable() :
+                    ImmutableArray.CreateRange(this);
+            }
+
+            internal ArrayBuilder<PendingBranch> GetAndRemoveBranches(LabelSymbol? label)
+            {
+                ArrayBuilder<PendingBranch>? result;
+                if (label is null)
+                {
+                    result = _branches;
+                    _branches = ArrayBuilder<PendingBranch>.GetInstance();
+                }
+                else if (_branchesToLabel is { } && _branchesToLabel.TryGetValue(label, out result))
+                {
+                    _branchesToLabel.Remove(label);
+                }
+                else
+                {
+                    result = ArrayBuilder<PendingBranch>.GetInstance();
+                }
+                return result;
+            }
+
+            internal void Add(PendingBranch branch)
+            {
+                var label = branch.Label;
+                if (label is null)
+                {
+                    _branches.Add(branch);
```

Comment:
In the case we have `Add(somePendingBranch)` called twice in a row we end up with duplicates in the collection. Is that expected as a part of this algoritthm?

Reply:
It looks like that is the expected outcome although that doesn't seem useful. (I attempted to keep the behavior unchanged where possible though.)